File: lib/backend/src/services/ExtractionService.ts
import fs from 'fs';
import path from 'path';
import { execFile } from 'child_process';
import { promisify } from 'util';

const execFileAsync = promisify(execFile);

/**
 * ExtractionService - Gerencia extração segura de arquivos PWR
 * Implementação baseada no Hytale F2P com:
 * - Staging directory para extrair sem corromper arquivos existentes
 * - Validação pré-extração (PWR exists, é readable, tamanho OK)
 * - Cleanup automático de directories temporários
 * - Detecção robusta de sucesso (client executable exists)
 */
export class ExtractionService {
  /**
   * Valida se arquivo PWR é válido antes de processar
   * @param pwrPath - Caminho do arquivo PWR
   * @returns true se válido, throws se inválido
   */
  static async validatePWRFile(pwrPath: string): Promise<boolean> {
    console.log(`[ExtractionService] Validating PWR file: ${pwrPath}`);

    // 1. Arquivo existe?
    if (!fs.existsSync(pwrPath)) {
      throw new Error(`PWR file not found: ${pwrPath}`);
    }

    // 2. É um arquivo (não diretório)?
    const stats = await fs.promises.stat(pwrPath);
    if (!stats.isFile()) {
      throw new Error(`PWR path is not a file: ${pwrPath}`);
    }

    // 3. Tamanho mínimo (PWR files devem ter pelo menos alguns MB)
    const minSize = 10 * 1024 * 1024; // 10MB minimum
    if (stats.size < minSize) {
      throw new Error(
        `PWR file suspiciously small: ${Math.round(stats.size / 1024 / 1024)}MB ` +
        `(expected at least ${Math.round(minSize / 1024 / 1024)}MB). ` +
        `File may be corrupted or incomplete.`
      );
    }

    // 4. É readable?
    try {
      await fs.promises.access(pwrPath, fs.constants.R_OK);
    } catch (err) {
      throw new Error(`PWR file is not readable: ${pwrPath}`);
    }

    console.log(`[ExtractionService] PWR validation passed. Size: ${Math.round(stats.size / 1024 / 1024)}MB`);
    return true;
  }

  /**
   * Extrai PWR usando Butler em directory temporário (staging)
   * Implementação segura que não corrompe game directory existente se falhar
   * 
   * @param pwrFile - Caminho do arquivo PWR validado
   * @param targetDir - Directory onde game será colocado após extração bem-sucedida
   * @param butlerPath - Caminho do executável butler
   * @param onProgress - Callback para progresso: (message, percent) => void
   * @returns true se extração bem-sucedida
   */
  static async extractPWR(
    pwrFile: string,
    targetDir: string,
    butlerPath: string,
    onProgress?: (message: string, percent?: number) => void
  ): Promise<boolean> {
    console.log(`[ExtractionService] Starting extraction of ${pwrFile} to ${targetDir}`);

    // Validar PWR antes de fazer qualquer coisa
    await this.validatePWRFile(pwrFile);

    // Validar butler existe
    if (!fs.existsSync(butlerPath)) {
      throw new Error(`Butler not found at: ${butlerPath}`);
    }

    const stagingDir = path.join(targetDir, 'staging-temp');
    let stagingCleanedUp = false;

    try {
      // 1. Preparar staging directory (limpar se existir)
      onProgress?.('Preparing staging directory...', 10);
      console.log(`[ExtractionService] Preparing staging dir: ${stagingDir}`);

      if (fs.existsSync(stagingDir)) {
        console.log(`[ExtractionService] Removing existing staging directory...`);
        await fs.promises.rm(stagingDir, { recursive: true, force: true });
      }
      await fs.promises.mkdir(stagingDir, { recursive: true });
      console.log(`[ExtractionService] Staging directory ready`);

      // 2. Verificar se game já está instalado no target
      // Se sim, skip (não precisa re-extrair)
      const existingClient = await this.findClientPath(targetDir);
      if (existingClient && fs.existsSync(existingClient)) {
        console.log(`[ExtractionService] Game already installed at ${targetDir}, skipping extraction`);
        onProgress?.('Game already installed, skipping extraction', 100);

        // Cleanup staging anyway
        if (fs.existsSync(stagingDir)) {
          await fs.promises.rm(stagingDir, { recursive: true, force: true });
          stagingCleanedUp = true;
        }

        return true;
      }

      // 3. Criar target directory se não existir
      if (!fs.existsSync(targetDir)) {
        console.log(`[ExtractionService] Creating target directory: ${targetDir}`);
        await fs.promises.mkdir(targetDir, { recursive: true });
      }

      // 4. Executar butler apply
      onProgress?.('Extracting game files...', 30);
      console.log(`[ExtractionService] Running butler with:`);
      console.log(`  - Butler: ${butlerPath}`);
      console.log(`  - PWR: ${pwrFile}`);
      console.log(`  - Staging: ${stagingDir}`);
      console.log(`  - Target: ${targetDir}`);

      const { stdout, stderr } = await execFileAsync(butlerPath, [
        'apply',
        '--staging-dir',
        stagingDir,
        pwrFile,
        targetDir
      ], {
        maxBuffer: 10 * 1024 * 1024, // 10MB buffer
        timeout: 10 * 60 * 1000 // 10 minutos timeout
      });

      console.log(`[ExtractionService] Butler stdout: ${stdout}`);
      if (stderr) {
        console.log(`[ExtractionService] Butler stderr: ${stderr}`);
      }

      // 5. Validar que extração foi bem-sucedida
      onProgress?.('Validating extraction...', 80);
      const extractedClient = await this.findClientPath(targetDir);
      if (!extractedClient) {
        throw new Error(
          'Extraction completed but client executable not found in target directory. ' +
          'This may indicate a corrupt PWR file or incomplete extraction.'
        );
      }

      const clientStats = await fs.promises.stat(extractedClient);
      if (clientStats.size < 20 * 1024 * 1024) { // 20MB minimum for game executable
        throw new Error(
          `Client executable suspiciously small: ${Math.round(clientStats.size / 1024 / 1024)}MB. ` +
          'Extraction may have failed or PWR file may be corrupt.'
        );
      }

      console.log(
        `[ExtractionService] Extraction validated successfully. ` +
        `Client found: ${extractedClient} (${Math.round(clientStats.size / 1024 / 1024)}MB)`
      );

      onProgress?.('Extraction completed successfully', 95);
      return true;
    } catch (error: any) {
      console.error(`[ExtractionService] Extraction failed: ${error.message}`);

      // Não deixar game directory quebrado
      if (fs.existsSync(targetDir) && !stagingCleanedUp) {
        try {
          const dir = await fs.promises.readdir(targetDir);
          // Se só tem staging, está seguro remover
          if (dir.length === 1 && dir[0] === 'staging-temp') {
            console.log(`[ExtractionService] Removing incomplete extraction from target...`);
            await fs.promises.rm(targetDir, { recursive: true, force: true });
          }
        } catch (cleanupErr) {
          console.warn(`[ExtractionService] Error during cleanup: ${cleanupErr}`);
        }
      }

      throw new Error(`Extraction failed: ${error.message}`);
    } finally {
      // SEMPRE cleanup staging directory mesmo se falhou
      if (fs.existsSync(stagingDir)) {
        try {
          console.log(`[ExtractionService] Cleaning up staging directory: ${stagingDir}`);
          await fs.promises.rm(stagingDir, { recursive: true, force: true });
        } catch (cleanupErr) {
          console.warn(`[ExtractionService] Warning: Could not cleanup staging directory: ${cleanupErr}`);
        }
      }
    }
  }

  /**
   * Procura por client executável em possíveis localizações
   * Retorna null se não encontrado
   */
  private static async findClientPath(gameDir: string): Promise<string | null> {
    const candidates = this.getClientCandidates(gameDir);

    for (const candidate of candidates) {
      if (fs.existsSync(candidate)) {
        try {
          const stats = await fs.promises.stat(candidate);
          if (stats.isFile()) {
            console.log(`[ExtractionService] Found client at: ${candidate}`);
            return candidate;
          }
        } catch (err) {
          console.warn(`[ExtractionService] Client candidate exists but not readable: ${candidate}`);
        }
      }
    }

    console.log(`[ExtractionService] Client not found in any candidate location`);
    return null;
  }

  /**
   * Retorna possíveis localizações do client executável por OS
   * Baseado no Hytale F2P - procura por HytaleClient ou similar em diferentes localizações
   */
  private static getClientCandidates(gameDir: string): string[] {
    const candidates: string[] = [];
    const platform = process.platform;

    // Windows - Procura por HytaleClient.exe
    if (platform === 'win32') {
      candidates.push(path.join(gameDir, 'Client', 'HytaleClient.exe'));
    }
    // macOS
    else if (platform === 'darwin') {
      candidates.push(path.join(gameDir, 'Client', 'Hytale.app', 'Contents', 'MacOS', 'HytaleClient'));
      candidates.push(path.join(gameDir, 'Client', 'HytaleClient'));
      candidates.push(path.join(gameDir, 'Hytale.app', 'Contents', 'MacOS', 'HytaleClient'));
    }
    // Linux
    else {
      candidates.push(path.join(gameDir, 'Client', 'HytaleClient'));
      candidates.push(path.join(gameDir, 'HytaleClient'));
    }

    return candidates;
  }

  /**
   * Verifica integridade de game directory após extração
   * @param gameDir - Directory do game extraído
   * @returns { valid: boolean, issues: string[] }
   */
  static async validateExtractedGame(gameDir: string): Promise<{ valid: boolean; issues: string[] }> {
    console.log(`[ExtractionService] Validating extracted game at: ${gameDir}`);
    const issues: string[] = [];

    // 1. Directory existe?
    if (!fs.existsSync(gameDir)) {
      return {
        valid: false,
        issues: [`Game directory does not exist: ${gameDir}`]
      };
    }

    // 2. Client executável existe e é válido?
    const clientPath = await this.findClientPath(gameDir);
    if (!clientPath) {
      issues.push('Client executable not found in game directory');
    } else {
      try {
        const stats = await fs.promises.stat(clientPath);
        if (stats.size < 20 * 1024 * 1024) {
          issues.push(
            `Client executable suspiciously small: ${Math.round(stats.size / 1024 / 1024)}MB ` +
            '(expected at least 20MB). Game may be corrupted.'
          );
        }
      } catch (err) {
        issues.push(`Cannot stat client executable: ${err}`);
      }
    }

    // 3. Client directory existe?
    const clientDir = path.join(gameDir, 'Client');
    if (!fs.existsSync(clientDir)) {
      issues.push(`Client directory missing: ${clientDir}`);
    }

    // 4. UserData pode ser criado?
    const userDataPath = path.join(clientDir, 'UserData');
    if (!fs.existsSync(userDataPath)) {
      try {
        await fs.promises.mkdir(userDataPath, { recursive: true });
        await fs.promises.rm(userDataPath, { recursive: true, force: true });
        console.log(`[ExtractionService] UserData directory will be created on first run`);
      } catch (err) {
        issues.push(`Cannot create UserData directory: ${err}`);
      }
    }

    const valid = issues.length === 0;
    console.log(`[ExtractionService] Validation result: ${valid ? 'VALID' : 'INVALID - ' + issues.join(', ')}`);

    return {
      valid,
      issues
    };
  }

  /**
   * Detecta se game está realmente instalado (não só se directory existe)
   * @param gameDir - Directory potencial do game
   * @returns true se client executável existe e parece válido
   */
  static async isGameFullyExtracted(gameDir: string): Promise<boolean> {
    if (!fs.existsSync(gameDir)) {
      return false;
    }

    try {
      const clientPath = await this.findClientPath(gameDir);
      if (!clientPath) {
        return false;
      }

      const stats = await fs.promises.stat(clientPath);
      const isValidSize = stats.size >= 50 * 1024 * 1024; // 50MB minimum

      console.log(
        `[ExtractionService] Game fully extracted check: ${isValidSize ? 'YES' : 'NO'} ` +
        `(client size: ${Math.round(stats.size / 1024 / 1024)}MB)`
      );

      return isValidSize;
    } catch (err) {
      console.log(`[ExtractionService] Game fully extracted check: NO (error: ${err})`);
      return false;
    }
  }

  /**
   * Remove directory extraído completamente (para repair/reinstall)
   */
  static async removeExtractedGame(gameDir: string): Promise<void> {
    if (fs.existsSync(gameDir)) {
      console.log(`[ExtractionService] Removing extracted game from: ${gameDir}`);
      await fs.promises.rm(gameDir, { recursive: true, force: true });
      console.log(`[ExtractionService] Game directory removed`);
    }
  }
}

File: lib/backend/src/services/ExtractionService.py
import os
import shutil
import subprocess
import platform
from .ButlerService import ButlerService

class ExtractionService:
    
    @staticmethod
    def validate_pwr_file(pwr_path):
        # print(f"[ExtractionService] Validating PWR file: {pwr_path}")

        if not os.path.exists(pwr_path):
            raise Exception(f"PWR file not found: {pwr_path}")

        if not os.path.isfile(pwr_path):
            raise Exception(f"PWR path is not a file: {pwr_path}")

        min_size = 10 * 1024 * 1024 # 10MB
        size = os.path.getsize(pwr_path)
        if size < min_size:
            raise Exception(f"PWR file suspiciously small: {size} bytes")

        # Check readability
        try:
            with open(pwr_path, 'rb') as f:
                f.read(1024)
        except Exception:
            raise Exception(f"PWR file is not readable: {pwr_path}")

        return True

    @staticmethod
    def extract_pwr(pwr_file, target_dir, butler_path, on_progress=None, skip_if_installed=True):
        # print(f"[ExtractionService] Starting extraction of {pwr_file} to {target_dir}")

        ExtractionService.validate_pwr_file(pwr_file)
        
        # Optionally skip extraction if client already exists (install-only scenario)
        if skip_if_installed:
            existing_client = ExtractionService.find_client_path(target_dir)
            if existing_client and os.path.exists(existing_client):
                try:
                    client_size = os.path.getsize(existing_client)
                except Exception:
                    client_size = 0
                if client_size < 20 * 1024 * 1024:
                    existing_client = None
                else:
                    pass
            if existing_client and os.path.exists(existing_client):
                 print(f"[ExtractionService] Game already installed at {target_dir}, skipping extraction")
                 if on_progress: on_progress("Game already installed, skipping extraction", 100)
                 
                 # Cleanup staging if exists
                 staging_dir = os.path.join(target_dir, 'staging-temp')
                 if os.path.exists(staging_dir):
                      shutil.rmtree(staging_dir, ignore_errors=True)
                 
                 return True

        staging_dir = os.path.join(target_dir, 'staging-temp')
        
        try:
            if on_progress: on_progress("Preparing staging directory...", 10)
            
            if os.path.exists(staging_dir):
                shutil.rmtree(staging_dir, ignore_errors=True)
            os.makedirs(staging_dir, exist_ok=True)
            
            # Create target directory if not exists
            os.makedirs(target_dir, exist_ok=True)

            if not os.path.exists(butler_path):
                 raise Exception(f"Butler not found at: {butler_path}")

            if on_progress: on_progress("Extracting files (Butler)...", 30)
            
            startupinfo = None
            if os.name == 'nt':
                startupinfo = subprocess.STARTUPINFO()
                startupinfo.dwFlags |= subprocess.STARTF_USESHOWWINDOW

            # Use Butler apply with --staging-dir as in the old backend
            # Command: butler apply --staging-dir <staging> <pwr> <target>
            cmd = [
                butler_path, 
                'apply', 
                '--staging-dir', 
                staging_dir, 
                pwr_file, 
                target_dir
            ]
            
            process = subprocess.Popen(
                cmd,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                startupinfo=startupinfo,
                text=True
            )
            
            try:
                stdout, stderr = process.communicate(timeout=600)
            except subprocess.TimeoutExpired:
                process.kill()
                stdout, stderr = process.communicate()
                raise Exception("Butler extraction timed out (10 minutes)")
            
            if process.returncode != 0:
                error_msg = (stderr or "").strip() or (stdout or "").strip() or "Unknown Butler error"
                raise Exception(f"Butler extraction failed: {error_msg}")
            
            if on_progress: on_progress("Finalizing installation...", 90)

            # Ensure executable permissions on Linux/Mac
            if platform.system() != 'Windows':
                ExtractionService._ensure_executable_permissions(target_dir)

            # Cleanup staging directory
            shutil.rmtree(staging_dir, ignore_errors=True)
            
            return True

        except Exception as e:
            print(f"Extraction failed: {e}")
            # Cleanup
            if os.path.exists(staging_dir):
                shutil.rmtree(staging_dir, ignore_errors=True)
            # Propagate the exception with message
            raise e

    @staticmethod
    def _ensure_executable_permissions(game_dir):
        import stat
        candidates = ExtractionService.get_client_candidates(game_dir)
        for candidate in candidates:
            if os.path.exists(candidate):
                try:
                    st = os.stat(candidate)
                    os.chmod(candidate, st.st_mode | stat.S_IEXEC)
                    # print(f"[ExtractionService] Set executable permission for: {candidate}")
                except Exception as e:
                    print(f"[ExtractionService] Failed to set executable permission for {candidate}: {e}")

    @staticmethod
    def validate_extracted_game(game_dir):
        # Check for key files
        # Windows: HytaleClient.exe or Client/HytaleClient.exe
        # We need to be flexible
        
        client_path = ExtractionService.find_client_path(game_dir)
        if client_path:
             return {"valid": True, "issues": []}

        return {"valid": False, "issues": ["Executable not found"]}

    @staticmethod
    def find_client_path(game_dir):
        candidates = ExtractionService.get_client_candidates(game_dir)
        for candidate in candidates:
            if os.path.exists(candidate) and os.path.isfile(candidate):
                # print(f"[ExtractionService] Found client at: {candidate}")
                return candidate
        return None

    @staticmethod
    def get_client_candidates(game_dir):
        candidates = []
        system = platform.system()
        if system == 'Windows':
            candidates.append(os.path.join(game_dir, 'Client', 'HytaleClient.exe'))
            candidates.append(os.path.join(game_dir, 'HytaleClient.exe')) # Fallback
            candidates.append(os.path.join(game_dir, 'Hytale', 'Client', 'HytaleClient.exe'))
        elif system == 'Darwin':
            candidates.append(os.path.join(game_dir, 'Client', 'Hytale.app', 'Contents', 'MacOS', 'HytaleClient'))
            candidates.append(os.path.join(game_dir, 'Client', 'HytaleClient'))
            candidates.append(os.path.join(game_dir, 'Hytale.app', 'Contents', 'MacOS', 'HytaleClient'))
        else:
            candidates.append(os.path.join(game_dir, 'Client', 'HytaleClient'))
            candidates.append(os.path.join(game_dir, 'HytaleClient'))
            
        return candidates
